### PR TITLE
[DI] Fix AutowirePass fatal error with classes that have non-existing parents

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 2.8 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18188 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by hason** [comment]
@nicolas-grekas You are a PHP hacker. Great!

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -242,12 +242,16 @@ private function getReflectionClass($id, Definition $definition)
         }
 
         $class = $this->container->getParameterBag()->resolveValue($class);
+        $throwingAutoloader = function () { throw new \ReflectionException(); };
+        $reflector = false;
 
         try {
+            spl_autoload_register($throwingAutoloader);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: this could even be outside the try/catch

**Comment by stof** [comment]
``$typeHint = $parameter->getClass()`` on line 79 could also trigger the loading of a class without a parent class. Could we apply a similar trick there (we will need to set a meaningful exception message though, saying that the class does not exist) ?

**Comment by nicolas-grekas** [comment]
@stof I wondered the same, but someone needs to give me a test case that would trigger the fatal error first :)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (lisachenko) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: @nicolas-grekas Very clever trick!

**Comment by stof** [comment]
@nicolas-grekas register a service for a class which has a constructor argument typehinted with a class which extend a missing class

**Comment by nicolas-grekas** [comment]
@stof done: https://github.com/symfony/symfony/pull/18600/files?w=1

**Comment by stof** [comment]
:+1:

**Comment by dunglas** [comment]
👍

**Comment by dunglas** [comment]
... and one more bug fixed, thank you @nicolas-grekas.

**Pull request merged**